pkg/hubble/recorder/pcap/pcap.go

```diff
@@ -0,0 +1,158 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pcap
+
+import (
+	"io"
+	"net"
+	"time"
+
+	"github.com/cilium/cilium/pkg/byteorder"
+)
+
+// Datalink defines the type of the first layer of the captured packet
+type Datalink uint32
+
+var (
+	Null     Datalink = 0
+	Ethernet Datalink = 1
+)
+
+const (
+	MagicNumber  = 0xa1b2c3d4
+	VersionMajor = 2
+	VersionMinor = 4
+
+	TimestampCorrection = 0
+	TimestampAccuracy   = 0
+
+	DefaultSnapLen = 65535
+)
+
+const (
+	GlobalHeaderByteLen = 24
+	RecordHeaderByteLen = 16
+)
+
+// Header contains the configurable parameters for the global pcap header
+type Header struct {
+	// SnapshotLength is the maximum length of captured packets
+	SnapshotLength uint32
+	// Datalink is the type of header at the beginning of the packet
+	Datalink Datalink
+}
+
+// Bytes returns the byte representation of the global pcap header
+func (h *Header) Bytes() []byte {
+	// https://wiki.wireshark.org/Development/LibpcapFileFormat#Global_Header
+	const (
+		offsetMagicNumber         = 0
+		offsetVersionMajor        = 4
+		offsetVersionMinor        = 6
+		offsetTimestampCorrection = 8
+		offsetTimestampAccuracy   = 12
+		offsetSnapshotLength      = 16
+		offsetDatalink            = 20
+	)
+
+	native := byteorder.Native
+	b := make([]byte, GlobalHeaderByteLen)
+
+	snapLen := h.SnapshotLength
+	if snapLen == 0 {
+		snapLen = DefaultSnapLen
+	}
+
+	native.PutUint32(b[offsetMagicNumber:], MagicNumber)
+	native.PutUint16(b[offsetVersionMajor:], VersionMajor)
+	native.PutUint16(b[offsetVersionMinor:], VersionMinor)
+	native.PutUint32(b[offsetTimestampCorrection:], TimestampCorrection)
+	native.PutUint32(b[offsetTimestampAccuracy:], TimestampAccuracy)
+	native.PutUint32(b[offsetSnapshotLength:], snapLen)
+	native.PutUint32(b[offsetDatalink:], uint32(h.Datalink))
+
+	return b
+}
+
+// Record contains the configurable parameters for a record header
+type Record struct {
+	// Timestamp is the time the packet was captured
+	Timestamp time.Time
+	// CaptureLength is the size of the captured packet
+	CaptureLength uint32
+	// OriginalLength is the size of the original packet
+	OriginalLength uint32
+}
+
+// Bytes returns the byte representation of the per-record pcap header
+func (r *Record) Bytes() []byte {
+	// https://wiki.wireshark.org/Development/LibpcapFileFormat#Record_.28Packet.29_Header
+	const (
+		offsetTimestampUnix   = 0
+		offsetTimestampMicros = 4
+		offsetCaptureLength   = 8
+		offsetOriginalLength  = 12
+	)
+
+	tsUnix := uint32(r.Timestamp.Unix())
+	tsMicros := uint32(r.Timestamp.Nanosecond() / 1000)
```

Comment:
Would be nicer to time.Truncate() here I think.

Reply:
internally truncate does `	return d - d%m`

Reply:
Yeah, I thought I could use it, but `time.Truncate()` does not do what I need here, I would still need to divide by 1000. `tsMicros` needs to be in the range of `[0, 1000000)`. Leaving as is.